packages/build-tools/src/android/gradle.ts

```diff
@@ -35,4 +37,57 @@ export async function runGradleCommand(
     },
     env: ctx.env,
   });
+  if (ctx.env.EAS_BUILD_RUNNER === 'eas-build' && process.platform === 'linux') {
+    adjustOOMScore(spawnPromise, ctx.logger);
+  }
+
+  await spawnPromise;
+}
+
+/**
+ * OOM Killer sometimes kills worker server while build is exceeding memory limits.
+ * `oom_score_adj` is a value between -1000 and 1000 and it defaults to 0.
+ * It defines which process is more likely to get killed (higher value more likely).
+ *
+ * This function sets oom_score_adj for Gradle process and all its child processes.
+ */
+function adjustOOMScore(spawnPromise: SpawnPromise<SpawnResult>, logger: bunyan): void {
+  setTimeout(
+    async () => {
+      try {
+        assert(spawnPromise.child.pid);
+        const children: number[] = [spawnPromise.child.pid];
+        let shouldRetry = true;
+        while (shouldRetry) {
+          const result = await spawn('pgrep', ['-P', children.join(',')], {
+            stdio: 'pipe',
+          });
+          const pids = result.stdout
+            .toString()
+            .split('\n')
+            .map((i) => Number(i.trim()))
+            .filter((i) => i);
+          shouldRetry = false;
+          for (const pid of pids) {
+            if (!children.includes(pid)) {
+              shouldRetry = true;
+              children.push(pid);
+            }
+          }
+        }
+        await Promise.all(
+          children.map(async (pid: number) => {
+            // Value 800 is just a guess here. It's probably higher than most other
+            // process. I didn't want to set it any higher, because I'm not sure if OOM Killer
+            // can start killing processes when there is still enough memory left.
+            await fs.writeFile(`/proc/${pid}/oom_score_adj`, '800\n');
```

Comment:
Also, as always (xD), I'd prefer to not use the magic number. Especially, there is \n after the value.